### PR TITLE
Implement Object operations for Spaces API

### PR DESCRIPTION
**Pull request by amoeba**
## Description

This implements the first half of the operations listed in https://github.com/sckott/analogsea/issues/140 which are the Spaces API operations related to Objects

## Related Issue
https://github.com/sckott/analogsea/issues/140

No tests yet, since all these tests require external calls. See discussion in https://github.com/sckott/analogsea/issues/140

### CODE REVIEW

File: R/spaces.R [hunk 1]
```diff
@@ -165,11 +165,276 @@ space_create <- function(name, spaces_key = NULL, spaces_secret = NULL, ...) {
...
+#' @param object (character) The Object to get
+#' @param space (character) The Space the Object is found in
+#' @param spaces_key (character) String containing a spaces access key. If
+#'   missing, defaults to value stored in an environment variable
+#'   \code{DO_SPACES_ACCESS_KEY}.
+#' @param spaces_secret (character) String containing the secret associated
+#'   with the spaces key. If missing, defaults to value stored in an environment
+#'   variable \code{DO_SPACES_SECRET_KEY}.
+#' @param ... Additional argument passed to \code{\link[aws.s3]{get_object}}
+#'
```
Comment (sckott) [inline R/spaces.R]: @amoeba if you want to include `@author <youremail>` tags for the functions you write go ahead

File: R/spaces.R [hunk 2]
```diff
@@ -165,11 +165,276 @@ space_create <- function(name, spaces_key = NULL, spaces_secret = NULL, ...) {
...
+#' @examples
+#' \dontrun{
+#' spaces_put_object("./some-file.txt", "my-object", "my-space")
+#'
+#' # Let the function choose an Object key automatically based on the file's
+#' # basename
+#' spaces_put_object("./some-file.txt", space = "my-space")
+#' }
+spaces_put_object <- function(file, object = basename(file), space, spaces_key = NULL, spaces_secret = NULL, ...) {
```
Comment (sckott) [inline R/spaces.R]: this function doesn't error well if the file doesn't exist - maybe an internal check for that?

File: R/spaces.R [hunk 3]
```diff
@@ -165,11 +165,276 @@ space_create <- function(name, spaces_key = NULL, spaces_secret = NULL, ...) {
...
+#'
+#' @return The raw response body of the request
+#' @export
+#'
+#' @references \url{https://developers.digitalocean.com/documentation/spaces/#get-object}
+#'
+#' @examples
+#' \dontrun{
+#' spaces_put_object("./some-file.txt", "my-object", "my-space")
+#' spaces_get_object("my-object", "my-space")
```
Comment (sckott) [inline R/spaces.R]: adding an example of how to convert the raw bytes to text would be good, e.g, `﻿rawToChar(x)`

File: R/spaces.R [hunk 4]
```diff
@@ -165,11 +165,276 @@ space_create <- function(name, spaces_key = NULL, spaces_secret = NULL, ...) {
...
+#' @export
+#'
+#' @references \url{https://developers.digitalocean.com/documentation/spaces/#get-object-info}
+#'
+#' @examples
+#' \dontrun{
+#' spaces_put_object("./some-file.txt", "my-object", "my-space")
+#' spaces_head_object("my-object", "my-space")
+#' }
+spaces_head_object <- function(object, space, spaces_key = NULL, spaces_secret = NULL, ...) {
```
Comment (sckott) [inline R/spaces.R]: Toughts on the output structure from this function?  Maybe make it into a named list instead? just via `attributes(x)`

File: R/spaces.R (hunk 1 above)
Comment (amoeba) [reply R/spaces.R]: Fair enough. I'm okay not doing this.

File: R/spaces.R (hunk 3 above)
Comment (amoeba) [reply R/spaces.R]: Great suggestion! Will do.

File: R/spaces.R (hunk 2 above)
Comment (amoeba) [reply R/spaces.R]: Roger.

File: R/spaces.R (hunk 4 above)
Comment (amoeba) [reply R/spaces.R]: I like that.

**Comment by amoeba** [comment]
Thanks for the comments. Will update this PR.

**Comment by sckott** [comment]
@amoeba sorry for delay - getting to looking this over hopefully today

File: DESCRIPTION [hunk 5]
```diff
@@ -10,6 +10,8 @@ Authors@R: c(
     person("Hadley", "Wickham", role = "aut", email = "hadley@rstudio.com"),
     person("Winston", "Chang", role = "aut", email = "winston@stdout.org"),
     person("Bob", "Rudis", role = "ctb", email = "bob@rudis.net"),
+    person("Bryce", "Mecum", role = "ctb", email = "brycemecum@gmail.com",
+           comment = c("ORCID" = "0000-0002-0381-3766")),
```
Comment (sckott) [inline DESCRIPTION]: nice, will gather ORCIDs for others later

File: R/spaces_acl_get.R [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+  response <- aws.s3::get_acl(object,
+                              space,
+                              check_region = FALSE,
+                              key = spaces_key,
+                              secret = spaces_secret,
+                              base_url = spaces_base,
+                              ...)
+
+  # if (!requireNamespace("xml2")) {
```
Comment (sckott) [inline R/spaces_acl_get.R]: @amoeba i forget, did you mean to leave this block commented out?

**Comment by sckott** [comment]
@amoeba is `space_create` working for you?

I'm getting an error on every try, checked that my env vars are there. 

SO with verbose curl output i'm seeing

```r
space_create('new_space_name', config=httr::verbose())
-> PUT /new_space_name/ HTTP/1.1
-> Host: nyc3.digitaloceanspaces.com
-> User-Agent: libcurl/7.54.0 r-curl/3.1 httr/1.3.1
... cut out some stuff
<- HTTP/1.1 400 Bad Request
<- Date: Thu, 25 Jan 2018 02:37:06 GMT
<- Content-Length: 153
<- Content-Type: text/plain; charset=utf-8
```

so the URL seems like `nyc3.digitaloceanspaces.com/space-name`

whereas in creating one in the DO platform, and the [API docs say the same](https://developers.digitalocean.com/documentation/spaces/#create-a-bucket), looks like the URL should be of the form

`space-name.nyc3.digitaloceanspaces.com`

thoughts? does it work for you?

File: R/spaces_acl_put.R [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#' spaces/#set-object-acls}
+#'
+#' @examples
+#' \dontrun{
+#' # First, create a Space and upload an Object to it
+#' space_create("my-space")
+#' spaces_object_put("some-file.txt", space = "my-space")
+#'
+#' # You can get a copy of the ACL
+#' acl <- spaces_acl_get("some-file.txt", "my-space") %>% read_xml()
```
Comment (sckott) [inline R/spaces_acl_put.R]: in examples, with Suggested pkgs, should use them conditionally on whether user has the pkg. 

so 

```r
if (requireNamespace("xml2")) {
  do the thing
}
```

and inside that block (if user has xml2 installed) namespace `xml2` calls

File: R/spaces_acl_put.R [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+#'
+#' # Modify it as needed, then update it
+#' spaces_acl_put("my-object", "my-space", acl)
```
Comment (sckott) [inline R/spaces_acl_put.R]: is `my-object` supposed to be equivalent to `some-file.txt` ?  if so could you change that, so peole can actually do the eg

**Comment by sckott** [comment]
I couldn't get `spaces_object_delete` to work. I kept getting a `The specified bucket does not exist` error, even though the bucket defiinitely existed. work for you?

File: R/spaces_acl_get.R (hunk 6 above)
Comment (amoeba) [reply R/spaces_acl_get.R]: Whoops. I did not! Will fix.

File: R/spaces_acl_put.R (hunk 7 above)
Comment (amoeba) [reply R/spaces_acl_put.R]: Oh, interesting. I can certainly do that.

File: R/spaces_acl_put.R (hunk 8 above)
Comment (amoeba) [reply R/spaces_acl_put.R]: That was a mistake. Fixed.

**Comment by amoeba** [comment]
Thanks for taking a look. I really appreciated the letting us have a go at this and the helpful comments.

I don't think I have ever been able to `space_create` working. I assumed it was just still in progress. But I think `spaces_object_delete` was working fine. I'll test now.

**Comment by amoeba** [comment]
Hrm, `spaces_object_delete` worked fine for me just now:

```r
> spaces_object_delete("GDP.csv", "test-analogsea")
[1] TRUE
```

**Comment by sckott** [comment]
Glad delete works. Wonder if they changed how spaces create works since that fxn was originally made here?

**Comment by amoeba** [comment]
I wonder. I'll take a look at it.

**Comment by amoeba** [comment]
Looks like a limitation of `aws.s3`. In `setup_s3_url`, which appears to generate the URLs used in `put_bucket` (which creates new S3 buckets), it hard-codes this logic:

```r
# handle S3-compatible storage URLs
    if (base_url != "s3.amazonaws.com") {
        if (isTRUE(verbose) && url_style != "path") {
            message("Non-AWS base URL requested. Switching to path-style URLs.")
        }
...
```

which looks like the cause of our problem here.

**Comment by sckott** [comment]
ah, that could be it, right, i assume then it adds the bucket/space name after the base url

**Comment by amoeba** [comment]
yup

**Comment by sckott** [comment]
bump @amoeba , are you waiting on me to do anything here?

**Comment by amoeba** [comment]
Hey @sckott thanks for the bump. This can be merged if you think so. I looked over all of the comments here and in the [related issue](https://github.com/sckott/analogsea/issues/140) and I think it's good to go.

**Comment by sckott** [comment]
okay, will take another quick look

**Comment by sckott** [comment]
@amoeba  can you add `man-roxygen` to `.Rbuildignore`

**Comment by amoeba** [comment]
Done!

**Comment by sckott** [comment]
thanks for the rbuildignore fix. 

hmm, did you see my comment about `space_create` not working? https://github.com/sckott/analogsea/pull/146#issuecomment-360343859

**Comment by amoeba** [comment]
I did. I was thinking that, since this PR is just for the Object API operations that I/we'd work on that elsewhere. It'll require either an upstream `aws.s3` fix or moving away from using `aws.s3`. Is that okay?

**Comment by sckott** [comment]
right, forgot about that. yep, we'll deal with fixing create later.

**Pull request merged**